src/api/spec/models/workflow/step/link_package_step_spec.rb

```diff
@@ -0,0 +1,305 @@
+require 'rails_helper'
+
+RSpec.describe Workflow::Step::LinkPackageStep, vcr: true do
+  let!(:user) { create(:confirmed_user, :with_home, login: 'Iggy') }
+  let(:token) { create(:workflow_token, user: user) }
+
+  subject do
+    described_class.new(step_instructions: step_instructions,
+                        scm_extractor_payload: scm_extractor_payload,
+                        token: token)
+  end
+
+  RSpec.shared_context 'source_project not provided' do
+    let(:step_instructions) { { source_package: package.name } }
+
+    it { expect { subject.call }.to(change(Package, :count).by(0)) }
+  end
+
+  RSpec.shared_context 'source_package not provided' do
+    let(:step_instructions) { { source_project: package.project.name } }
+
+    it { expect { subject.call }.to(change(Package, :count).by(0)) }
+  end
+
+  RSpec.shared_context 'failed when source_package does not exist' do
+    let(:step_instructions) do
+      {
+        source_project: project.name,
+        source_package: 'this_package_does_not_exist'
+      }
+    end
+
+    it { expect { subject.call }.to raise_error(Workflow::Step::Errors::CanNotLinkPackageNotFound) }
+  end
+
+  RSpec.shared_context 'project and package does not exist' do
+    let(:step_instructions) do
+      {
+        source_project: 'invalid project',
+        source_package: 'invalid package'
+      }
+    end
+
+    it { expect { subject.call }.to raise_error(Project::Errors::UnknownObjectError) }
+  end
+
+  RSpec.shared_context 'target package already exists' do
+    # Emulates that the target project and package already existed
+    let!(:linked_project) { create(:project, name: "home:#{user.login}:#{package.project.name}:PR-1", maintainer: user) }
+    let!(:linked_package) { create(:package_with_file, name: package.name, project: linked_project) }
+
+    let(:step_instructions) do
+      {
+        source_project: project.name,
+        source_package: package.name
+      }
+    end
+
+    it { expect { subject.call }.to raise_error(Workflow::Step::Errors::DuplicatePackageError) }
+  end
+
+  RSpec.shared_context 'failed without link permissions' do
+    let(:step_instructions) do
+      {
+        source_project: project.name,
+        source_package: package.name
+      }
+    end
+
+    before do
+      allow_any_instance_of(Package).to receive(:disabled_for?).with('sourceaccess', nil, nil).and_return(true)
```

Comment:
It's not possible to mock this directly without `any_instance`?

Reply:
That's what we did at the beginning, however, we didn't manage to make the test pass. After hours of trying different alternatives, we decided to go back to `allow_any_instance_of`and not to block this PR anymore.